source/common/network/win32_socket_handle_impl.cc

```diff
@@ -84,5 +117,65 @@ void Win32SocketHandleImpl::reEnableEventBasedOnIOResult(const Api::IoCallUint64
   }
 }
 
+Api::IoCallUint64Result Win32SocketHandleImpl::drainToPeekBuffer(size_t length) {
+  size_t total_bytes_read = 0;
+  while (peek_buffer_.length() < length) {
+    Buffer::Reservation reservation = peek_buffer_.reserveForRead();
+    Api::IoCallUint64Result result = IoSocketHandleImpl::readv(
+        reservation.length(), reservation.slices(), reservation.numSlices());
+    uint64_t bytes_to_commit = result.ok() ? result.return_value_ : 0;
+    reservation.commit(bytes_to_commit);
+    total_bytes_read += bytes_to_commit;
+    if (!result.ok() || bytes_to_commit == 0) {
+      return result;
+    }
+  }
+  return Api::IoCallUint64Result(total_bytes_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {}));
+}
+
+Api::IoCallUint64Result Win32SocketHandleImpl::readFromPeekBuffer(void* buffer, size_t length) {
+  uint64_t copy_size = std::min(peek_buffer_.length(), static_cast<uint64_t>(length));
```

Comment:
My mistake, the OwnedImpl::copyOut did the calucation of min length.  https://github.com/envoyproxy/envoy/blob/d79a3ab49f1aa522d0a465385425e3e00c8db147/source/common/buffer/buffer_impl.cc#L109

so we needn't calculate it here.